agents/host_manager/host_manager.py

```diff
@@ -0,0 +1,476 @@
+            items in the current database.  The target_state must be
+            'up' or 'down'.
+          reload_site_config (bool): Default is True.  If True, the
+            site config file is parsed in order to (re-)populate the
+            database of child Agent instances.
+
+        Examples:
+
+          ::
+
+            _update_target_states(session, reload_site_config=True)
+            _update_target_states(session, requests=[('thermo1', 'down')])
+            _update_target_states(session, requests=[('all', 'up')])
+
+        Notes:
+          First, the site config file is parsed and used to update the
+          internal database of child instances (unless
+          ``reload_site_config`` has been set to False).  Any
+          previously unknown child Agent is added to the internal
+          tracking database, and assigned a target_state of "down".
+          Any previously known child Agent instance is not modified in
+          the tracking database (unless a specific request is given,
+          through ``requests``).  If any child Agent instances in the
+          internal database appear to have been removed from the site
+          config, then they are set to have target_state "down" and
+          will be deleted from the database when that state is
+          reached.
+
+          State update requests in the ``requests`` list are processed
+          in order.  For example, if the requests were [('all', 'up'),
+          ('data1', 'down')].  This would result in setting all known
+          children to have target_state "up", except for "data1" which
+          would be given target state of "down".
+
+        """
+        if reload_site_config:
+            # Update the list of Agent instances.
+            agent_keys = yield self._get_instance_list()
+            session.add_message('Loaded %i agent instance configs.' %
+                                len(agent_keys))
+            downers = 0
+            for k, item in self.database.items():
+                # Any new keys?
+                if k not in agent_keys:
+                    item['retired'] = True
+                    item['target_state'] = 'down'
+                    downers += 1
+            if downers > 0:
+                session.add_message('Retiring %i instances that disappeared '
+                                    'from config.' % downers)
+            for k in agent_keys:
+                if not k in self.database:
+                    state = 'down'
+                    if k[0] == 'docker':
+                        agent_script = 'docker'
+                        prot = self._get_docker_helper(self.docker_services[k[1]])
+                        if prot.status[0] == None:
+                            session.add_message(
+                                'On startup, detected active container for %s' % k[1])
+                            ## Leave it up?  No.
+                            #state = 'up'
```

Comment:
Do these get left up? It appears if I startup a bunch of containers with an HM, kill the HM, then startup the HM again it detects these active containers, then requests launches for them and states they are launched.

```
2022-01-27T19:42:48+0000 startup-op: launching manager
2022-01-27T19:42:48+0000 start called for manager
2022-01-27T19:42:48+0000 manager:0 Status is now "starting".
2022-01-27T19:42:48+0000 manager:0 Status is now "running".
2022-01-27T19:42:49+0000 manager:0 Loaded 3 agent instance configs.
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for ocs-client
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for fake-data1
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for ocs-registry
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:ocs-client
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:fake-data1
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:ocs-registry
2022-01-27T19:42:52+0000 manager:0 Launched docker:ocs-client
2022-01-27T19:42:52+0000 manager:0 Launched docker:fake-data1
2022-01-27T19:42:52+0000 manager:0 Launched docker:ocs-registry
```

Reply:
Logging was misleading ... fixed in 0e7ac6.  Won't show the Requested and Launched lines.

This must have been with --initial-state=up?  Without that switch, HM would stop those containers.

Reply:
It was with `--initial-state=up`, yes.

Reply:
I believe this is resolved, let me know if not.

Reply:
Sorry, ran out of time last night. This looks good to me now.